### PR TITLE
[Feat] Migrate and add new NSA Decode operator(kernel)

### PR DESCRIPTION
**Pull request by wch0810**
## Description

Add Native Sparse Attention Decode (NSA Decode) kernel for Ascend NPU, implemented in `example_tilelang_nsa_decode.py` based on TileLang Developer mode (hybrid) — `pass_configs` four-True (combineCV + auto_cv_sync + auto_sync + memory_planning) with explicit L1/L0C/UB allocation and 5 GM workspace relay (auto-allocated via `workspace_idx=[6,7,8,9,10]`).

Single kernel (decode only, seq_q=1). Developer mode hybrid: compiler auto-splits Cube/Vector scopes, auto-inserts cross-core flag sync (no `T.barrier_all`, no manual `T.Scope`, no manual `set_flag/wait_flag`). Layered test suite (28 cases) covering L0 threshold / L1 functional (including DeepSeek-V3 typical config D=128, S=16, BS=64) / L2 exception / Boundary special-value.

Tested on Ascend 910B3, fp16, B=2, SEQ_LEN=64, H=1, HQ=16, D=16, S=1, block_size=32. Precision verified against PyTorch CPU golden (online softmax with neg-scaled max + axpy fusion, mixed-tolerance dual-gate per `precision-standard.md` §4.1: float16 atol=2⁻¹⁴, rtol=2⁻⁹, max_abs_limit=1e-1, required_ratio=0.99, 28/28 PASS).

## Golden Config (B=2, SEQ_LEN=64, H=1, HQ=16, D=16, S=1, BS=32, fp16)

| Kernel                       | Task Duration (us) | Block Dim | Mix Block Dim | Max Diff   |
| ---------------------------- | ------------------ | --------- | ------------- | ---------- |
| TileLang NSA Decode          | 8.66               | 2         | 4 (2C+2V)     | 4.883e-04  |

## Precision (float16, precision-standard.md §4.1)

| Level              | Cases | Result | matched_ratio       | max_abs                  |
| ------------------ | ----- | ------ | ------------------- | ------------------------ |
| L0 (blocking)      | 3     | ✅ PASS | 0.9922 ~ 0.9980     | 2.441e-04                |
| L1 (blocking)      | 17    | ✅ PASS | 0.9941 ~ 1.0000     | 0 ~ 4.883e-04            |
| L2 (non-blocking)  | 3     | ✅ PASS | — (correctly rejected) | —                     |
| Boundary (non-blocking) | 5 | ✅ PASS | 1.0000           | 0 ~ 2.441e-04            |

- L1 includes 2 model typical config cases: `typical_d128_s16` (B=1, seq=1024, D=128, S=16, BS=64) + `typical_d128_s8` (B=1, seq=512, D=128, S=8, BS=64)
- Thresholds (float16): atol=2⁻¹⁴ (6.10e-5), rtol=2⁻⁹ (1.95e-3), max_abs_limit=1e-1, required_ratio=0.99
- Dual-gate AND: matched_ratio ≥ 0.99 **AND** max_abs ≤ 1e-1
- INF/NAN structural compare (§3.1): positions must match, not counted in numeric tolerance

## File Structure (2 files in this PR)

```
examples/deepseek_nsa/example_tilelang_nsa_decode/
├── example_tilelang_nsa_decode.py      # kernel + self-contained CI smoke test (346 lines)
└── test_example_tilelang_nsa_decode.py  # golden + L0/L1/L2/Boundary precision tests (388 lines)
```

### CODE REVIEW

File: examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
+"""NSA Decode layered precision test suite: L0/L1/L2/Boundary.
+
+This file is the single precision-test entry for `example_tilelang_nsa_decode.py`.
+It embeds:
+  - ref_nsa_decode:     PyTorch CPU golden reference (moved from operator file).
+  - check_precision:    mixed-tolerance dual-gate check (precision-standard.md §4.1).
+  - test_nsa_l0/l1/l2/boundary: layered test cases (L0/L1 blocking, L2/Boundary non-blocking).
+  - main(--level):      unified dispatch + exit code.
+
+Performance profiling (do_bench + msprof) is in perf_example_tilelang_nsa_decode.py.
```
Comment (fengz72) [inline examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py]: 此文件不在此次pr中

File: examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        for h in range(H):
+            for s in range(bc[b, h].item()):
+                idx = block_indices[b, 0, h, s].item()
+                assert idx * block_size + block_size <= seq_len, (
+                    f"Valid block (b={b}, h={h}, s={s}) idx={idx} is incomplete: "
+                    f"idx*BS+BS={idx * block_size + block_size} > seq_len={seq_len}."
+                )
+
+
+def allocate_workspace(batch, head_kv, G, S, block_size, D, device="npu"):
```
Comment (fengz72) [inline examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py]: 可以使用workspace_idx, 让框架自动分配内存

File: examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    output = kernel(q_npu, k_npu, v_npu, ri_npu, bc_npu, ws0, ws1, ws2, ws3, ws4)
+    torch.npu.synchronize()
+    return output
+
+
+# =============================================================================
+# Naive PyTorch CPU reference (inline for CI smoke test only).
+# The full golden (ref_nsa_decode) lives in test_example_tilelang_nsa_decode.py.
+# =============================================================================
+def _naive_ref(q, k, v, block_indices, block_counts, block_size):
```
Comment (fengz72) [inline examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py]: kernel 文件 _naive_ref 与 test 文件 ref_nsa_decode 两份等价实现, 保留一份既可

File: examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+# L1 functional tests — blocking
+# =============================================================================
+def _gen_uniform(low, high):
+    def fn(shape):
+        return (torch.rand(shape, dtype=torch.float32) * (high - low) + low).to(torch.float16)
+
+    return fn
+
+
+def _gen_zeros(shape):
```
Comment (fengz72) [inline examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py]: _gen_zeros/_gen_inf/_gen_nan/_gen_dbound 定义未使用

**Review by fengz72 — COMMENTED** [review]
B, SEQ_LEN, H, HQ, D, S, BS 参数建议使用模型的典型值

File: examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py (hunk 1 above)
Comment (wch0810) [reply examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py]: 1.已在 L1 新增 2 个 DeepSeek-V3 NSA 论文典型配置用例：
- `typical_d128_s16`: B=1, seq=1024, D=128, S=16, BS=64（论文推荐配置）
- `typical_d128_s8`: B=1, seq=512, D=128, S=8, BS=64（中等规模配置）
两者精度均 PASS（matched_ratio=1.0000, max_abs=1.221e-04）。L0 baseline 保留 D=16, S=1, BS=32（与 GPU baseline 一致，便于对照精度）。

2.已删除 test 文件 docstring 中对 `perf_example_tilelang_nsa_decode.py` 的引用。

File: examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py (hunk 2 above)
Comment (wch0810) [reply examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py]: 已改为 `@tilelang.jit(out_idx=[5], workspace_idx=[6, 7, 8, 9, 10], pass_configs=PASS_CONFIGS)`，删除 `allocate_workspace` 函数，调用方简化为 `kernel(q, k, v, ri, bc)`（不再手动传 5 个 workspace tensor）。性能提升 -4.63%（min 9.08µs → 8.66µs）。

File: examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py (hunk 3 above)
Comment (wch0810) [reply examples/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py]: 已删除 `_naive_ref` 函数，

File: examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py (hunk 4 above)
Comment (wch0810) [reply examples/deepseek_nsa/example_tilelang_nsa_decode/test_example_tilelang_nsa_decode.py]: 已直接删除 4 个未使用的函数

**Review by fengz72 — DISMISSED** [review]

File: examples_experiment/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        workspace_4: [block_num, S*BS, D]     float16 (gathered V, V→C)
+    """
+    assert kv_head_num == 1, "kv_head_num must be 1"
+    G = query_heads // kv_head_num
+    assert G >= 16, f"G={G} must be >= 16 for L0C fractal"
+    assert G % 2 == 0, f"G={G} must be even for vid split"
+
+    dtype = "float16"
+    accum_dtype = "float32"
+    BS = block_size
```
Comment (ChaoyangJi) [inline examples_experiment/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py:57]: 显式约束block size的硬件对齐要求

下面的实现同时使用BS/2`做两个id的行切分、BS/∥8分配packed mask,而T.tile.compare'要求输入总字节数按256B对齐。因此当前实现实际至少要求block_size%8==0。传入非8的倍数时可能在AscendC lowering阶段直接失败；在不执行同样检查的后端上，mask容量不足或最后一行被遗漏。请在入口处增加明确断言。

Comment (wch0810) [reply examples_experiment/deepseek_nsa/example_tilelang_nsa_decode/example_tilelang_nsa_decode.py]: 已改进，添加断言
